Source/WebCore/dom/SpatialImageControls.cpp

```diff
@@ -0,0 +1,219 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SpatialImageControls.h"
+
+#include "ElementInlines.h"
+#include "ElementRareData.h"
+#include "Event.h"
+#include "EventLoop.h"
+#include "HTMLButtonElement.h"
+#include "HTMLDivElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLNames.h"
+#include "HTMLStyleElement.h"
+#include "MouseEvent.h"
+#include "RenderImage.h"
+#include "ShadowRoot.h"
+#include "TreeScopeInlines.h"
+#include "UserAgentStyleSheets.h"
+#include <wtf/text/AtomString.h>
+
+namespace WebCore {
+namespace SpatialImageControls {
+
+#if ENABLE(SPATIAL_IMAGE_CONTROLS)
+
+static const AtomString& spatialImageControlsElementIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("spatial-image-controls"_s);
+    return identifier;
+}
+
+static const AtomString& spatialImageControlsButtonIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("spatial-image-controls-button"_s);
+    return identifier;
+}
+
+bool hasSpatialImageControls(const HTMLElement& element)
+{
+    RefPtr shadowRoot = element.shadowRoot();
+    if (!shadowRoot || shadowRoot->mode() != ShadowRootMode::UserAgent)
+        return false;
+
+    return shadowRoot->hasElementWithId(spatialImageControlsElementIdentifier());
+}
+
+static RefPtr<HTMLElement> spatialImageControlsHost(const Node& node)
+{
+    RefPtr host = dynamicDowncast<HTMLElement>(node.shadowHost());
+    if (!host)
+        return nullptr;
+
+    return hasSpatialImageControls(*host) ? host : nullptr;
+}
+
+bool isSpatialImageControlsButtonElement(const Element& element)
+{
+    return spatialImageControlsHost(element) && element.getIdAttribute() == spatialImageControlsButtonIdentifier();
+}
+
+bool shouldHaveSpatialControls(HTMLImageElement& element)
+{
+    if (!element.document().settings().spatialImageControlsEnabled())
+        return false;
+
+    bool hasSpatialcontrolsAttribute = element.hasAttributeWithoutSynchronization(HTMLNames::controlsAttr);
```

Comment:
Is this spec'ed at all?

Reply:
Not at this time, which is why it's behind an experimental feature flag. That would of course have to change before this ever would become on by default.